packages/app-store/mercadopago/lib/PaymentService.ts

```diff
@@ -0,0 +1,198 @@
+  private mercadoPago: MercadoPago;
+
+  constructor(credentials: { id: number; key: Prisma.JsonValue }) {
+    const parse = mercadoPagoCredentialSchema.safeParse(credentials.key);
+    if (!parse.success) {
+      throw new Error("Invalid `credentials`");
+    }
+
+    this.mercadoPago = new MercadoPago({
+      clientId: process.env.MERCADOPAGO_CLIENT_ID || "",
+      clientSecret: process.env.MERCADOPAGO_CLIENT_SECRET || "",
+      userCredentials: {
+        id: credentials.id,
+        key: parse.data,
+      },
+    });
+  }
+
+  private async getPayment(where: Prisma.PaymentWhereInput) {
+    const payment = await prisma.payment.findFirst({ where });
+    if (!payment) throw new Error("Payment not found");
+    if (!payment.externalId) throw new Error("Payment externalId not found");
+    if (!payment.data) throw new Error("Payment `data` not set");
+    return {
+      ...payment,
+      externalId: payment.externalId,
+      mercadoPagoPaymentId: (payment.data as { paymentId: string }).paymentId,
+    };
+  }
+
+  async create(
+    payment: Pick<Prisma.PaymentUncheckedCreateInput, "amount" | "currency">,
+    bookingId: Booking["id"],
+    bookerEmail: string,
+    paymentOption: PaymentOption,
+    eventTitle?: string
+  ) {
+    try {
+      // Ensure that the payment service can support the passed payment option
+      if (paymentOptionEnum.parse(paymentOption) !== "ON_BOOKING") {
+        throw new Error("Payment option is not compatible with create method");
+      }
+
+      const booking = await prisma.booking.findFirst({
+        select: {
+          uid: true,
+          title: true,
+          eventTypeId: true,
+        },
+        where: {
+          id: bookingId,
+        },
+      });
+
+      if (!booking) {
+        throw new Error("Booking not found");
+      }
+      if (!booking.eventTypeId) {
+        throw new Error("Booking not related to an EventType");
+      }
+
+      const paymentUid = uuidv4();
+
+      const preferenceResult = await this.mercadoPago.createPreference({
+        amount: payment.amount / 100,
+        currency: payment.currency,
+        paymentUid,
+        bookingId,
+        eventTypeId: booking.eventTypeId,
+        bookerEmail,
+        eventName: eventTitle || "",
+        notificationUrl: `${WEBAPP_URL}/api/integrations/mercadopago/webhook?external_reference=${paymentUid}`,
+        returnUrl: `${WEBAPP_URL}/booking/${booking.uid}?mercadoPagoPaymentStatus=success`,
+        cancelUrl: `${WEBAPP_URL}/payment/${paymentUid}`,
+      });
+
+      const paymentData = await prisma.payment.create({
+        data: {
+          uid: paymentUid,
+          app: {
+            connect: {
+              slug: "mercadopago",
+            },
+          },
+          booking: {
+            connect: {
+              id: bookingId,
+            },
+          },
+          amount: payment.amount,
+          externalId: preferenceResult.id,
```

Comment:
Here we don't have access to the MercadoPago (MP) `Payment` yet. The preference lets the user create a payment when redirected to the MP checkout. It will forward the metadata and the Cal.com `uid`.